src/libraries/Microsoft.PowerFx.Interpreter/EvalVisitor.cs

```diff
@@ -85,30 +103,34 @@ public override FormulaValue Visit(TableNode node, SymbolContext context)
             return tableValue;
         }
 
-        public override FormulaValue Visit(RecordNode node, SymbolContext context)
+        public override async ValueTask<FormulaValue> Visit(RecordNode node, SymbolContext context)
         {
             var fields = new List<NamedValue>();
 
             foreach (var field in node.Fields)
             {
+                CheckCancel();
+
                 var name = field.Key;
                 var value = field.Value;
 
-                var rhsValue = value.Accept(this, context);
+                var rhsValue = await value.Accept(this, context);
                 fields.Add(new NamedValue(name.Value, rhsValue));
             }
 
             return new InMemoryRecordValue(node.IRContext, fields);
         }
 
-        public override FormulaValue Visit(LazyEvalNode node, SymbolContext context)
+        public override async ValueTask<FormulaValue> Visit(LazyEvalNode node, SymbolContext context)
         {
-            var val = node.Child.Accept(this, context);
+            var val = await node.Child.Accept(this, context);
             return val;
         }
 
-        public override FormulaValue Visit(CallNode node, SymbolContext context)
+        public override async ValueTask<FormulaValue> Visit(CallNode node, SymbolContext context)
         {
+            CheckCancel();
```

Comment:
It feels like pattern for CheckCancel is more of an art rather than a pattern. Here we are calling it at the beginning of the method. In other places we are not calling it at all, and then we also call it in the loop.

Maybe we can agree that every visitor method needs to call it at the beginning. Then most loops are also covered (as the call Accept inside). ?